### PR TITLE
pkg/lolcat-rs

### PR DESCRIPTION
**Pull request by mirrorwitch**
## Description of changes

Rust implementation of lolcat for speed (in my machine I get something like 26x faster results).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/lolcat-rs/default.nix [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{ lib, fetchgit, rustPlatform }:
+
+rustPlatform.buildRustPackage rec {
+  pname = "lolcat-rs";
+  version = "1.5.0";
+
+  src = fetchgit {
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/lolcat-rs/default.nix:7]: This should use fetchFromGitHub

Comment (mirrorwitch) [reply pkgs/tools/misc/lolcat-rs/default.nix:7]: I tried that first, but it failed because the upstream author didn't create any releases or tarballs.

Comment (xfix) [reply pkgs/tools/misc/lolcat-rs/default.nix:7]: I think in this specific case `fetchCrate` may make more sense, because the upstream doesn't seem to tag releases on GitHub.

```nix
  src = fetchCrate {
    crateName = "lolcat";
    inherit version;
    hash = "sha256-RVYmBfF9BlS9WSBcnptlgZ2nF7b7NC+58cZMHVX7SxQ=";
  };
```

**Comment by donovanglover** [comment]
Result of `nixpkgs-review pr 258479` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lolcat-rs</li>
  </ul>
</details>

File: pkgs/tools/misc/lolcat-rs/default.nix [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    hash = "sha256-RVYmBfF9BlS9WSBcnptlgZ2nF7b7NC+58cZMHVX7SxQ=";
+  };
+
+  cargoHash = "sha256-hfcQsB4c+0/BmAV7Q0NUcGjedc35teDGRv5fDyyvp2c=";
+
+  meta = with lib; {
+    description = "A rainbow version of cat, implemented in Rust for speed";
+    homepage = "https://github.com/ur0/lolcat";
+    license = licenses.mit;
+    maintainers = with maintainers; [ mirrorwitch ];
```
Comment (donovanglover) [inline pkgs/tools/misc/lolcat-rs/default.nix:19]:
```suggestion
    maintainers = with maintainers; [ mirrorwitch ];
    mainProgram = "lolcat";
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -9961,6 +9961,8 @@ with pkgs;
 
   lolcat = callPackage ../tools/misc/lolcat { };
 
+  lolcat-rs = callPackage ../tools/misc/lolcat-rs { };
```
Comment (donovanglover) [inline pkgs/top-level/all-packages.nix:9964]: Use the new [`pkgs/by-name`](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name) format so you don't have to edit this file

**Review by donovanglover — CHANGES_REQUESTED** [review]

**Comment by donovanglover** [comment]
Would be great if you can squash the commits so there's one for the maintainer and one for the package

**Comment by Artturin** [comment]
There already is a faster lolcat https://github.com/NixOS/nixpkgs/blob/fe0b3b663e98c85db7f08ab3a4ac318c523c0684/pkgs/tools/misc/clolcat/default.nix it's repo hasn't been updated since 2018

There is also one rust lolcat already https://github.com/NixOS/nixpkgs/blob/fe0b3b663e98c85db7f08ab3a4ac318c523c0684/pkgs/tools/misc/dotacat/default.nix


Please discuss with their maintainers @felipeqq2 and @traxys  to decide which one should be kept to not have  4 novelty program lolcats

**Comment by xfix** [comment]
I disagree about having too many lolcats. https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md says the following:

> How realistic is it that it will be used by other people? It's good that nixpkgs caters to various niches, but if it's a niche of 5 people it's probably too small.

https://github.com/ur0/lolcat repository on GitHub has 87 stars, and https://lib.rs/crates/lolcat counts 3 downloads per day from crates.io. It's also packaged on AUR by two different people: https://aur.archlinux.org/packages/rust-lolcat-git and https://aur.archlinux.org/packages/lolcat-rs.

I don't think this specific package is too niche for nixpkgs, personally.

**Comment by Artturin** [comment]
Not too niche but it's a novelty program and do we need 4 lolcat implementations

**Comment by mirrorwitch** [comment]
By nature, lolcat is human interaction software (I like using it in my shell prompt and motd).  Latency matters for interaction because it produces annoyance; the experimental limit for perception is as low as 2ms, with 10ms response latency getting annoying for gamers, and 100ms for human beings.[1]  With the original ruby lolcat, `sh -c 'echo "hello world" | lolcat` on a 2022 Thinkpad T14s on a fresh NixOS install takes 105ms, making it completely unusable, likely to send the system into trashing, and a significant contributor to global warming.[2]

Having commited to adding cutting-edge 1989 technology to my terminals,[3] I decided to take upon the task of benchmarking various lolcat rust implementations.  To test performance in a slower system as well as compilation outside x86, I used my Raspberry Pi 4 (Armbian).  Because I think for typical lolcat use, latency to repeated operation matters more than throughput, but to also add some heft, I did the following test:

    for i in $(seq 1 500); do cat /usr/share/common-licenses/GPL-3 | $lolcat ; done

(A tad more complicated than that cos some of them need "--force" options when outputing to a nonterminal.)  Results:

    cat:         1.09user 1.39system 0:01.38elapsed 179%CPU (0avgtext+0avgdata 1700maxresident)k
    lolcat-rs:   43.98user 65.34system 1:52.65elapsed 97%CPU (0avgtext+0avgdata 2488maxresident)k
    clolcat:     1.65user 2.46system 0:03.21elapsed 128%CPU (0avgtext+0avgdata 2124maxresident)k
    lcat:        14.46user 7.20system 0:24.29elapsed 89%CPU (0avgtext+0avgdata 2548maxresident)k
    dotacat:     42.52user 7.71system 0:53.40elapsed 94%CPU (0avgtext+0avgdata 3624maxresident)k
    lolcrab:     14.51user 6.79system 0:23.08elapsed 92%CPU (0avgtext+0avgdata 2504maxresident)k
    lolcat:      2.21e+08user 6e+07system 3679230:13.12elapsed 90%CPU (0avgtext+0avgdata 14588maxresident)k)[4]

Somewhat disappointingly to my Rust fangirl ass, not only did lolcat-rs performed poorly, but it seems like no one can even approach clolcat for speed (due to better concurrency?).  In fact it's almost spooky how close clolcat comes to regular cat; I had to triple-check it wasn't turning off colorization due to the pipe.

Another feature to consider in a drop-in replacement is option parity.  Do the options match the original lolcat?

    lolcat-rs    no
    clolcat      partially
    lcat         no
    dotacat      partially
    lolcrab      partially

Other considerations:
 - lcat features better colour spaces and Unicode support.
 - lolcrab forks lcat to add random noise patterns.  It also seems to be the only one to implement the original's animation mode.
 - By downloads and recency on crates.io, lcat wins, lolcrab second.

In conclusion: I'm closing my own PR; I don't think lolcat-rs is worth it when it brings nothing new to the table, and clolcat soundly beats it on every regard as a drop-in replacement.  I'm not sold on dotacat either, to be honest.  There's a good argument to be made for lcat, in my opinion, since it has better colours and compability, and is both popular and active.

1: https://danluu.com/input-lag/
2: for clarity: this is hyperbole for the sake of humour (as is this whole overcommitment to the bit, I don't care about lolcat all that much).
3: 24-bit colour.
4: Actually it was like, 22 minutes.
